File: benchmarks/tasks/matmul/Matmul.java
// Three nested loops over flat arrays of doubles: the shape of numerical code,
// and of nothing else. `i k j` order, so the innermost loop walks in order.

public class Matmul {
    static final int N = 256;

    public static void main(String[] args) {
        double[] a = new double[N * N];
        double[] b = new double[N * N];
        double[] c = new double[N * N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                a[i * N + j] = (i + j) % 10;
                b[i * N + j] = (i * j) % 10;
            }
        }
        for (int i = 0; i < N; i++) {
            for (int k = 0; k < N; k++) {
                double aik = a[i * N + k];
                for (int j = 0; j < N; j++) {
                    c[i * N + j] += aik * b[k * N + j];
                }
            }
        }
        double total = 0;
        for (int i = 0; i < N; i++) total += c[i * N + i];
        System.out.println((long) total);
    }
}

File: benchmarks/tasks/matmul/matmul.rs
// Three nested loops over flat arrays of doubles: the shape of numerical code,
// and of nothing else. `i k j` order, so the innermost loop walks in order.

fn main() {
    const N: usize = 256;
    let mut a = vec![0.0f64; N * N];
    let mut b = vec![0.0f64; N * N];
    let mut c = vec![0.0f64; N * N];
    for i in 0..N {
        for j in 0..N {
            a[i * N + j] = ((i + j) % 10) as f64;
            b[i * N + j] = ((i * j) % 10) as f64;
        }
    }
    for i in 0..N {
        for k in 0..N {
            let aik = a[i * N + k];
            for j in 0..N {
                c[i * N + j] += aik * b[k * N + j];
            }
        }
    }
    let total: f64 = (0..N).map(|i| c[i * N + i]).sum();
    println!("{}", total as i64);
}
